research/conditional/moe_layers/expert_choice.py

```diff
@@ -410,92 +203,22 @@ def gating_postprocess_bmm(self, x, topk_values, one_hot, weight):
-            x, topk_indices, batch_size, self.lin1_weight
-        )
-        x = F.relu(x)
-        x = self.gating_postprocess_onehot_with_linear(
-            x, topk_values, one_hot, self.lin2_weight
-        )
-        return x
-
-    def full_bmm(
-        self, x: torch.Tensor, topk_indices: torch.Tensor, topk_values, batch_size
-    ):
-        x, one_hot = self.extract_with_linear_bmm(
-            x, topk_indices, batch_size, self.lin1_weight
-        )
-        with measure_time(self, "activation"):
-            x = F.relu(x)
-        x = self.gating_postprocess_bmm(x, topk_values, one_hot, self.lin2_weight)
+        assert x.shape == (batch_size, seq_len, self.doutput)
         return x
 
-    def extract_chosen_tokens_select(
-        self, x: torch.Tensor, topk, topk_indices: torch.Tensor, batch_size
-    ):
-        # flatten x s. t. first dimension is tokens instead of batch_size x seq_len
-        with measure_time(self, "first_flatten"):
-            x = x.flatten(start_dim=0, end_dim=1)
-        with measure_time(self, "index_select"):
-            x = torch.index_select(x, dim=0, index=topk_indices.flatten())
-        with measure_time(self, "reshape"):
-            x = x.reshape((self.n_experts, topk, self.dmodel))
-        return x, None
-
-    def feed_forward(self, x: torch.Tensor, topk: int) -> torch.Tensor:
-        # feed through ff
-        with measure_time(self, "ff"):
-            # lin1 maps from (n_experts, topk, dmodel) to (n_experts, topk, exp_size)
-            x = einsum(
-                "n_exp topk dmodel, n_exp dmodel exp_size -> n_exp topk exp_size",
-                x,
-                self.lin1_weight,
-            )
-            x = F.relu(x)
-
-            # lin2 maps from (n_experts, topk, exp_size) to (n_experts, topk, dmodel)
-            x = einsum(
-                "n_exp topk exp_size, n_exp exp_size doutput -> n_exp topk doutput",
-                x,
-                self.lin2_weight,
-            )
-            ash.assert_shape("e k m", x, e=self.n_experts, k=topk, m=self.doutput)
-        return x
-
-    def gating_postprocess_onehot(
-        self, x, batch_size, topk, seq_len, topk_values, topk_indices, one_hot
-    ):
-        topk //= seq_len
-        with measure_time(self, "multiply_softmax"):
-            x = x.reshape(self.n_experts, topk, seq_len, self.dmodel)
-            x = einsum(
-                "n_exp topk seq_len dmodel, n_exp topk seq_len, n_exp topk seq_len batch_size "
-                "-> batch_size seq_len dmodel",
-                x,
-                topk_values,
-                one_hot,
-            )
-        return x
-
-    def gating_postprocess_select(
+    def merge_index_select(
         self, x, batch_size, topk, seq_len, topk_values, topk_indices, one_hot
     ):
         # multiply by softmax
         with measure_time(self, "multiply_softmax"):
             ash.assert_shape("e k", topk_values, e=self.n_experts, k=topk)
```

Comment:
możemy to wywalić przy okazji?

Reply:
niee, index select jest optymalny przy dużych modelach i dużych Granularity, ta implementacja jakbyśmy trenowali coś dużego jest domyślna.